File: comm/convex_hull.cpp
/*
 * Convex hull.
 * Verified by https://beta.atcoder.jp/contests/agc021/submissions/3179522
 * Dependencies: algorithm, complex, vector
 */
typedef ll coord; // the type of coordinates
typedef complex<coord> P;
namespace std {
  bool operator<(const P& a, const P& b) {
    return real(a) != real(b) ? real(a) < real(b) : imag(a) < imag(b);
  }
}

coord det(P a, P b) {
  return imag(conj(a) * b);
}

vector<P> convex_hull(vector<P> ps) {
  int n = ps.size(), k = 0;
  sort(ps.begin(), ps.end());
  vector<P> ch(2 * n);
  // lower
  for (int i = 0; i < n; k++, i++) {
    while (k >= 2 && det(ps[i] - ch[k - 2], ch[k - 1] - ch[k - 2]) >= 0) --k;
    ch[k] = ps[i];
  }
  // upper
  for (int i = n-2, t = k+1; i >= 0; k++, i--) {
    while (k >= t && det(ps[i] - ch[k - 2], ch[k - 1] - ch[k - 2]) >= 0) --k;
    ch[k] = ps[i];
  }
  ch.resize(k-1);
  return ch;
}

File: comm/convex_hull.rs
// Convex hull.
// Returns the points of the convex hull in the counter-clockwise order.
// Verified by: AGC021 B
// https://atcoder.jp/contests/agc021/submissions/22697344
// QUPC 2014 G
// https://atcoder.jp/contests/qupc2014/submissions/23493247
type Coord = i64; // the type of coordinates
type P = (Coord, Coord);

fn det((ax, ay): P, (bx, by): P) -> Coord {
    ax * by - ay * bx
}

fn sub((ax, ay): P, (bx, by): P) -> P {
    (ax - bx, ay - by)
}

fn convex_hull(ps: &[P]) -> Vec<P> {
    let n = ps.len();
    if n == 0 {
        return vec![];
    }
    let mut k = 0;
    let mut ps = ps.to_vec();
    ps.sort();
    let mut ch = vec![(0, 0); 2 * n];
    // lower
    for i in 0..n {
        while k >= 2 &&
            det(sub(ps[i], ch[k - 2]), sub(ch[k - 1], ch[k - 2])) >= 0 {
            k -= 1;
        }
        ch[k] = ps[i];
        k += 1;
    }
    // upper
    let t = k + 1;
    for i in (0..n - 1).rev() {
        while k >= t &&
            det(sub(ps[i], ch[k - 2]), sub(ch[k - 1], ch[k - 2])) >= 0 {
            k -= 1;
        }
        ch[k] = ps[i];
        k += 1;
    }
    ch.truncate(k - 1);
    ch
}
